src/docfx/build/xref/XrefResolver.cs

```diff
@@ -93,41 +93,50 @@ internal class XrefResolver
             return (null, resolvedHref, display, xrefSpec?.DeclaringFile);
         }
 
-        public (Error?, ExternalXrefSpec?) ResolveXrefSpec(SourceInfo<string> uid, Document referencingFile)
+        public (Error?, ExternalXrefSpec?, string? href) ResolveXrefSpec(SourceInfo<string> uid, Document referencingFile)
```

Comment:
If this method returns `href` and `IXrefSpec` (without calling `ToExternal`), would line 62 be changed to call this method instead and avoid duplicating URL processing logic?
```suggestion
        public (Error?, IXrefSpec?, string? href) ResolveXrefSpec(SourceInfo<string> uid, Document referencingFile)
```

Reply:
@yufeih  No, that will make break xref spec-level circular reference.
ResolveXrefSpec will visit all xref properties.
One way to avoid duplicate href resolve would be move it inside 
`private (Error?, IXrefSpec?) Resolve(SourceInfo<string> uid, Document referencingFile)`

Reply:
If `ResolveXrefSpec` returns `IXrefSpec` _without calling `ToExternalXrefSpec`_, then the method won't expand all xref properties?

Reply:
@yufeih Without calling ToExternalXrefSpec will lead to xrefProperties lost for internal specs.

Reply:
I've moved shared linkBuilder logic inside Resolve.